README.md
# GitHub Actions for Maintenance Windows

This repo contains custom [GitHub Actions](https://docs.github.com/en/actions) that enable support
for starting and stopping maintenance windows within our [Site24x7](https://www.site24x7.com/)
uptime monitoring, so that any downtime during a deployment is marked as having occurred during
maintenance.

## Start maintenance

Use this action to start a maintenance window.

### Parameters

| Parameter | Description  |
| --------- | ------------ |
| `app-url` | See [below](#common-connection-parameters). |
| `app-key` | See [below](#common-connection-parameters). |
| `monitor-id` | The ID of the monitor that should be suspended during the maintenance window. This will be a long numeric string such as `633320000127123456` |
| `duration` | The planned duration of the maintenance window, in minutes. The window will automatically end after this period. |

### Outputs

| Output           | Description  |
| ---------------- | ------------ |
| `maintenance-id` | The ID of the maintenance window instance that was created. This can be used to cancel the maintenance window if required. |

### Example

```yaml
uses: richteambs/maintainizr-action/start
with:
    app-url: ${{ secrets.MAINTZ_FUNC_URL }}
    app-key: ${{ secrets.MAINTZ_FUNC_APP_KEY }}
    monitor-id: '12345'
    duration: 1 # 1 minute
```

## End maintenance

Use this action to end a maintenance window.

If you don't call this action, the window will automatically end after the `duration` specified when
the window was started.

If you call this action when the window is already ended, it will _not_ report an error. This means
you can safely call it even if you have overrun the maintenance window, or if you have called it
multiple times.

If you pass an invalid maintenance id, the action will report an error.

### Parameters

| Parameter | Description  |
| --------- | ------------ |
| `app-url` | See [below](#common-connection-parameters). |
| `app-key` | See [below](#common-connection-parameters). |
| `maintenance-id` | The ID of the maintenance window to terminate. This should be the value output by the `start` action. |

### Example

```yaml
uses: richteambs/maintainizr-action/end
with:
    app-url: '${{ secrets.MAINTZ_FUNC_URL }}'
    app-key: ${{ secrets.MAINTZ_FUNC_APP_KEY }}
    maintenance-id: ${{ needs.start_maintenance.outputs.maintenance-id }}
```

## Common connection parameters

The actions in this repo work by contacting the Maintainizr API, which in turn contacts the Site24x7
API (the Maintainizr API provides a simplified facade). You therefore need to provide some
parameters that specify the connection details for the Maintainizr API. These parameters are common
to all the actions in this repository.

| Parameter | Description  |
| --------- | ------------ |
| `app-url` | The root URL of the Maintainizr API. |
| `app-key` | A secret key that is required for security purposes. <br/>:warning: This value should not be stored in plaintext in your repo or anywhere else in GitHub, except as an [encrypted secret](https://docs.github.com/en/actions/security-guides/encrypted-secrets). |

[This
article](https://mbs.atlassian.net/wiki/spaces/DO/pages/2419523603/Creating+an+Azure+Function+App+Service+Connection/)
documents how to find out the values to use for these parameters.

## Contributing

Information on how you can contribute to this repo is [here](./docs/CONTRIBUTING.md).

Developer documentation is [here](./docs/DevelopingAndReleasing.md).

## Credits

[Rocket Icon Vectors by Vecteezy](https://www.vecteezy.com/free-vector/rocket-icon)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

### Changed

- Update to latest NPM packages

### Fixed

- Corrected developer documentation around automatically moving the major tag

## [0.2.0] - 2022-03-28

### Added

- Use PAT instead of GITHUB\_TOKEN to enable triggering workflows when a PR is created

## [0.1.0] - 2022-03-28

### Added

- Initial version

[0.2.0]: https://github.com/richteambs/maintainizr-action/compare/v0.1.0...v0.2.0

[0.1.0]: https://github.com/richteambs/maintainizr-action/releases/tag/v0.1.0


tests/main.test.ts
import { wait } from '../src/wait';
import * as process from 'process';
import * as cp from 'child_process';
import * as path from 'path';
import { expect, test } from '@jest/globals';

test('throws invalid number', async () => {
  const input = parseInt('foo', 10);
  await expect(wait(input)).rejects.toThrow('milliseconds not a number');
});

test('wait 500 ms', async () => {
  const start = new Date();
  await wait(500);
  const end = new Date();
  const delta = Math.abs(end.getTime() - start.getTime());
  expect(delta).toBeGreaterThan(450);
});

// shows how the runner will run a javascript action with env / stdout protocol
test('test runs', () => {
  process.env['INPUT_MILLISECONDS'] = '500';
  const np = process.execPath;
  const ip = path.join(__dirname, '..', 'lib', 'main.js');
  const options: cp.ExecFileSyncOptions = {
    env: process.env,
  };
  console.log(cp.execFileSync(np, [ip], options).toString());
});


src/__tests__/end.test.ts
import * as path from 'path';
import { expect, describe, it } from '@jest/globals';
import { AxiosResponse } from 'axios';
import { ProblemDetails } from '../maintainizr-api';
import { invokeActionScript, getAllErrors, ActionResult } from './test-utils';

interface EndActionParams {
  appUrl: string | null;
  appKey: string | null;
  maintenanceId: string | null;
  apiResponse: AxiosResponse<never | ProblemDetails> | null;
}

const DefaultParams: EndActionParams = {
  appUrl: 'http://dummy.example.com/',
  appKey: '12345',
  maintenanceId: '987654321',
  apiResponse: null,
};

describe('gh end action', () => {
  it('succeeds when the API returns a good response', () => {
    const fakeApiResponse: AxiosResponse = {
      status: 200,
      statusText: 'OK',
      headers: {},
      config: {},
      data: {
        maintenanceId: '902010',
      },
    };

    const params = { ...DefaultParams };
    params.apiResponse = fakeApiResponse;
    const result = runAction(params);

    expect(result.isError).toBeFalsy();
    expect(getAllErrors(result).length).toBe(0);
  });

  it('fails when the API returns a bad response', () => {
    const fakeApiResponse: AxiosResponse<ProblemDetails> = {
      status: 502,
      statusText: 'Bad gateway',
      headers: {},
      config: {},
      data: {
        status: 502,
        type: 'https://tools.ietf.org/html/rfc7231#section-6.6.3',
      },
    };

    const params = { ...DefaultParams };
    params.apiResponse = fakeApiResponse;
    const result = runAction(params);

    expect(result.isError).toBeTruthy();
    expect(getAllErrors(result)).toContain('Call failed to Maintainizr API: 502 Bad gateway');
  });

  it('errors when maintenance-id is not set', () => {
    const params = { ...DefaultParams };
    params.maintenanceId = null;
    const result = runAction(params);

    expect(result.isError).toBeTruthy();
    expect(getAllErrors(result)).toContain('Input required and not supplied: maintenance-id');
  });

  it('errors when maintenance-id is empty', () => {
    const params = { ...DefaultParams };
    params.maintenanceId = '';
    const result = runAction(params);

    expect(result.isError).toBeTruthy();
    expect(getAllErrors(result)).toContain('Input required and not supplied: maintenance-id');
  });
});

function runAction(params: EndActionParams): ActionResult {
  const env: NodeJS.ProcessEnv = {
    MAINTZ_USE_FAKE_API: 'true',
  };

  if (params.appUrl !== null) {
    env['INPUT_APP-URL'] = params.appUrl;
  }

  if (params.appKey !== null) {
    env['INPUT_APP-KEY'] = params.appKey;
  }

  if (params.maintenanceId !== null) {
    env['INPUT_MAINTENANCE-ID'] = params.maintenanceId;
  }

  if (params.apiResponse !== null) {
    env['MAINTZ_FAKE_RESPONSE_CONTENT'] = JSON.stringify(params.apiResponse);
  }

  return invokeActionScript(path.join(__dirname, '../../lib/end.js'), env);
}


src/__tests__/start.test.ts
import * as path from 'path';
import { expect, describe, it } from '@jest/globals';
import { AxiosResponse } from 'axios';
import { ProblemDetails, MaintenanceOccurrence } from '../maintainizr-api';
import { invokeActionScript, getAllErrors, getOutputVariables, ActionResult } from './test-utils';

interface StartActionParams {
  appUrl: string | null;
  appKey: string | null;
  monitorId: string | null;
  duration: number | null;
  apiResponse: AxiosResponse<MaintenanceOccurrence | ProblemDetails> | null;
}

const DefaultParams: StartActionParams = {
  appUrl: 'http://dummy.example.com/',
  appKey: '12345',
  monitorId: '987654321',
  duration: 10,
  apiResponse: null,
};

describe('gh start action', () => {
  it('succeeds and sets the output when the API returns a good response', () => {
    const fakeApiResponse: AxiosResponse<MaintenanceOccurrence> = {
      status: 200,
      statusText: 'OK',
      headers: {},
      config: {},
      data: {
        maintenanceId: '902010',
        displayName: 'Ad hoc maintenance',
        status: 'scheduled',
        startDateTime: '2022-02-28T09:38:00+00:00',
        endDateTime: '2022-02-28T09:39:00+00:00',
        monitors: ['1234567890'],
      },
    };

    const params = { ...DefaultParams };
    params.apiResponse = fakeApiResponse;
    const result = runAction(params);

    expect(result.isError).toBeFalsy();
    expect(getAllErrors(result).length).toBe(0);
    const o = getOutputVariables(result);
    expect(o).toHaveProperty('maintenance-id', '902010');
  });

  it('fails when the API returns a bad response', () => {
    const fakeApiResponse: AxiosResponse<ProblemDetails> = {
      status: 502,
      statusText: 'Bad gateway',
      headers: {},
      config: {},
      data: {
        status: 502,
        type: 'https://tools.ietf.org/html/rfc7231#section-6.6.3',
      },
    };

    const params = { ...DefaultParams };
    params.apiResponse = fakeApiResponse;
    const result = runAction(params);

    expect(result.isError).toBeTruthy();
    expect(getAllErrors(result)).toContain('Call failed to Maintainizr API: 502 Bad gateway');
  });

  it('errors when monitor-id is not set', () => {
    const params = { ...DefaultParams };
    params.monitorId = null;
    const result = runAction(params);

    expect(result.isError).toBeTruthy();
    expect(getAllErrors(result)).toContain('Input required and not supplied: monitor-id');
  });

  it('errors when monitor-id is empty', () => {
    const params = { ...DefaultParams };
    params.monitorId = '';
    const result = runAction(params);

    expect(result.isError).toBeTruthy();
    expect(getAllErrors(result)).toContain('Input required and not supplied: monitor-id');
  });

  it('errors when duration is not set', () => {
    const params = { ...DefaultParams };
    params.duration = null;
    const result = runAction(params);

    expect(result.isError).toBeTruthy();
    expect(getAllErrors(result)).toContain('Input required and not supplied: duration');
  });

  it('errors when duration is zero', () => {
    const params = { ...DefaultParams };
    params.duration = 0;
    const result = runAction(params);

    expect(result.isError).toBeTruthy();
    expect(getAllErrors(result)).toContain("Input 'duration' must be between 1 and 1440 minutes");
  });
});

function runAction(params: StartActionParams): ActionResult {
  const env: NodeJS.ProcessEnv = {
    MAINTZ_USE_FAKE_API: 'true',
  };

  if (params.appUrl !== null) {
    env['INPUT_APP-URL'] = params.appUrl;
  }

  if (params.appKey !== null) {
    env['INPUT_APP-KEY'] = params.appKey;
  }

  if (params.monitorId !== null) {
    env['INPUT_MONITOR-ID'] = params.monitorId;
  }

  if (params.duration !== null) {
    env['INPUT_DURATION'] = `${params.duration}`;
  }

  if (params.apiResponse !== null) {
    env['MAINTZ_FAKE_RESPONSE_CONTENT'] = JSON.stringify(params.apiResponse);
  }

  return invokeActionScript(path.join(__dirname, '../../lib/start.js'), env);
}


src/end.ts
import * as core from '@actions/core';
import axios, { AxiosError } from 'axios';

import { MaintainizrApi } from './maintainizr-api';

async function run(): Promise<void> {
  try {
    const appUrl: string = core.getInput('app-url', { required: true });
    const appKey: string = core.getInput('app-key', { required: true });
    const maintenanceID: string = core.getInput('maintenance-id', { required: true });

    const api = new MaintainizrApi(appUrl, appKey);
    const response = await api.endMaintenance(maintenanceID);

    core.startGroup('Maintainizr API response');
    core.info(`status: ${response.status} ${response.statusText}`);
    if (response.data) {
      core.info(JSON.stringify(response.data));
    } else {
      core.info('No response content');
    }
    core.endGroup();

    if (response.status < 200 || response.status > 299) {
      core.setFailed(`Call failed to Maintainizr API: ${response.status} ${response.statusText}`);
    }
  } catch (err) {
    const error = err as Error | AxiosError;
    if (axios.isAxiosError(error)) {
      core.startGroup('Error details');
      core.info(JSON.stringify(error.toJSON()));
      core.endGroup();
    }

    core.setFailed(error.message);
  }
}

void run();


.eslintrc.js
module.exports = {
  extends: ['prettier'],
  plugins: ['@typescript-eslint', 'prettier', 'jest', 'eslint-plugin-import'],
  parser: '@typescript-eslint/parser',
  settings: {
    'import/parsers': {
      '@typescript-eslint/parser': ['.ts', '.tsx'],
    },
    'import/resolver': {
      typescript: {},
    },
  },
  rules: {
    'prettier/prettier': ['error', { endOfLine: 'auto' }],
    '@typescript-eslint/indent': ['error', 2],
    'linebreak-style': ['error', require('os').EOL === '\r\n' ? 'windows' : 'unix'],
    'max-len': ['error', { code: 120, comments: 120 }],
    'object-curly-newline': 0,
  },
  overrides: [
    {
      files: ['*.ts'],
      extends: [
        'plugin:@typescript-eslint/recommended',
        'plugin:@typescript-eslint/recommended-requiring-type-checking',
        'prettier',
      ],
      parserOptions: {
        ecmaVersion: 9,
        sourceType: 'module',
        // 👇 Note that we use a separate tsconfig just for linting, so that we include the test code
        project: './tsconfig.eslint.json',
      },
      rules: {
        'no-unused-vars': 'off',
        '@typescript-eslint/no-unused-vars': ['error', { argsIgnorePattern: '^_' }],
        'import/no-extraneous-dependencies': [
          'error',
          {
            devDependencies: ['**/*.test.ts', '**/tests/**/*.ts'],
          },
        ],
        'no-underscore-dangle': ['error', { allowAfterThis: true }],
        'import/extensions': ['error', { ts: 'ignorePackages' }],
        'no-plusplus': 'off',
        'no-continue': 'off',
      },
    },
    {
      files: ['*.test.ts'],
      extends: [
        'plugin:@typescript-eslint/recommended',
        'plugin:@typescript-eslint/recommended-requiring-type-checking',
        'prettier',
      ],
      rules: {
        'prefer-arrow-callback': 0,
        'func-names': 0,
        '@typescript-eslint/no-non-null-assertion': 0,
        'no-unused-expressions': 0,
        '@typescript-eslint/unbound-method': 'off',
        'jest/unbound-method': 'error',
      },
    },
  ],
  env: {
    node: true,
    es6: true,
    'jest/globals': true,
  },
};


src/maintainizr-api.ts
import * as core from '@actions/core';
import axios, { AxiosError, AxiosRequestConfig, AxiosRequestHeaders, AxiosResponse } from 'axios';

interface StartMaintenanceParams {
  duration: number;
}

export interface MaintenanceOccurrence {
  maintenanceId: string;
  displayName: string;
  status: string;
  startDateTime: string;
  endDateTime: string;
  monitors: string[];
}

// https://datatracker.ietf.org/doc/html/rfc7807
export interface ProblemDetails {
  type?: string;
  title?: string;
  detail?: string;
  status: number;
  instance?: string;
}

type StartMaintenanceResponseContent = MaintenanceOccurrence | ProblemDetails;

type EndMaintenanceResponseContent = never | ProblemDetails;

export class MaintainizrApi {
  rootUrl: string;
  appKey: string;

  constructor(rootUrl: string, appKey: string) {
    // Strip any trailing slashes to normalize
    this.rootUrl = rootUrl.replace(/\/+$/g, '');
    this.appKey = appKey;
  }

  async startMaintenance(monitorID: string, duration: number): Promise<AxiosResponse<StartMaintenanceResponseContent>> {
    if (process.env['MAINTZ_USE_FAKE_API']?.toLowerCase() === 'true') {
      return this.callFakeStartMaintenanceApi(monitorID, duration);
    } else {
      return this.callRealStartMaintenanceApi(monitorID, duration);
    }
  }

  public endMaintenance(maintenanceID: string): Promise<AxiosResponse<EndMaintenanceResponseContent>> {
    if (process.env['MAINTZ_USE_FAKE_API']?.toLowerCase() === 'true') {
      return this.callFakeEndMaintenanceApi(maintenanceID);
    } else {
      return this.callRealEndMaintenanceApi(maintenanceID);
    }
  }

  private async callRealStartMaintenanceApi(
    monitorID: string,
    duration: number
  ): Promise<AxiosResponse<StartMaintenanceResponseContent>> {
    try {
      const url = `${this.rootUrl}/api/monitors/${monitorID}/start-maintenance`;
      core.info(`Target URL: ${url}`);

      const body = {
        durationMinutes: duration,
      };
      const headers: AxiosRequestHeaders = { 'x-functions-key': this.appKey };
      const config: AxiosRequestConfig = {
        headers,
        // 👇 Don't throw an error for non-success HTTP status, we'll handle it ourselves
        validateStatus: () => true,
      };

      const response = await axios.post<StartMaintenanceParams, AxiosResponse<MaintenanceOccurrence | ProblemDetails>>(
        url,
        body,
        config
      );

      return response;
    } catch (err) {
      const error = err as Error | AxiosError;
      if (axios.isAxiosError(error)) {
        core.startGroup('Axios error details');
        core.info(JSON.stringify(error.toJSON()));
        core.endGroup();
      }

      throw err;
    }
  }

  private async callRealEndMaintenanceApi(
    maintenanceID: string
  ): Promise<AxiosResponse<EndMaintenanceResponseContent>> {
    try {
      const url = `${this.rootUrl}/api/maintenance/${maintenanceID}/end`;
      core.info(`Target URL: ${url}`);

      const headers: AxiosRequestHeaders = { 'x-functions-key': this.appKey };
      const config: AxiosRequestConfig = {
        headers,
        // 👇 Don't throw an error for non-success HTTP status, we'll handle it ourselves
        validateStatus: () => true,
      };

      const response = await axios.post<StartMaintenanceParams, AxiosResponse<EndMaintenanceResponseContent>>(
        url,
        null,
        config
      );

      return response;
    } catch (err) {
      const error = err as Error | AxiosError;
      if (axios.isAxiosError(error)) {
        core.startGroup('Axios error details');
        core.info(JSON.stringify(error.toJSON()));
        core.endGroup();
      }

      throw err;
    }
  }

  private async callFakeStartMaintenanceApi(
    _monitorID: string,
    _duration: number
  ): Promise<AxiosResponse<StartMaintenanceResponseContent>> {
    const responseContent = process.env['MAINTZ_FAKE_RESPONSE_CONTENT'];
    if (!responseContent) {
      throw new Error('Environment variable MAINTZ_USE_FAKE_API is set, but MAINTZ_FAKE_RESPONSE_CONTENT is not set');
    }

    const response: AxiosResponse<StartMaintenanceResponseContent> = <AxiosResponse<StartMaintenanceResponseContent>>(
      JSON.parse(responseContent)
    );
    return Promise.resolve(response);
  }

  private callFakeEndMaintenanceApi(_maintenanceID: string): Promise<AxiosResponse<EndMaintenanceResponseContent>> {
    const responseContent = process.env['MAINTZ_FAKE_RESPONSE_CONTENT'];
    if (!responseContent) {
      throw new Error('Environment variable MAINTZ_USE_FAKE_API is set, but MAINTZ_FAKE_RESPONSE_CONTENT is not set');
    }

    const response: AxiosResponse<EndMaintenanceResponseContent> = <AxiosResponse<EndMaintenanceResponseContent>>(
      JSON.parse(responseContent)
    );
    return Promise.resolve(response);
  }
}

export function isMaintenanceOccurrence(data: unknown): data is MaintenanceOccurrence {
  return (data as MaintenanceOccurrence).maintenanceId !== undefined;
}


src/start.ts
import * as core from '@actions/core';
import axios, { AxiosError } from 'axios';

import { MaintainizrApi, isMaintenanceOccurrence } from './maintainizr-api';

async function run(): Promise<void> {
  try {
    const appUrl: string = core.getInput('app-url', { required: true });
    const appKey: string = core.getInput('app-key', { required: true });
    const monitorID: string = core.getInput('monitor-id', { required: true });
    const duration: number = parseInt(core.getInput('duration', { required: true }));

    if (duration < 1 || duration > 1440) {
      core.setFailed("Input 'duration' must be between 1 and 1440 minutes");
      return;
    }

    const api = new MaintainizrApi(appUrl, appKey);
    const response = await api.startMaintenance(monitorID, duration);

    core.startGroup('Maintainizr API response');
    core.info(`status: ${response.status} ${response.statusText}`);
    if (response.data) {
      core.info(JSON.stringify(response.data));
    } else {
      core.info('No response content');
    }
    core.endGroup();

    if (response.status < 200 || response.status > 299 || !isMaintenanceOccurrence(response.data)) {
      core.setFailed(`Call failed to Maintainizr API: ${response.status} ${response.statusText}`);
    } else {
      core.setOutput('maintenance-id', response.data.maintenanceId);
    }
  } catch (err) {
    const error = err as Error | AxiosError;
    if (axios.isAxiosError(error)) {
      core.startGroup('Error details');
      core.info(JSON.stringify(error.toJSON()));
      core.endGroup();
    }

    core.setFailed(error.message);
  }
}

void run();
